de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/project/FileSystemChangeListener.java

```diff
@@ -1,529 +1,867 @@
+        if (newBaseName != null) {
+            relativePath = IntelliJPathImpl.fromString(newBaseName)
+                .append(relativePath.removeFirstSegments(1));
+        }
 
-        if (incomingFilesToFilterFor.remove(newFile)) {
-            return;
+        if (newFileName != null) {
+            relativePath = relativePath.removeLastSegments(1)
+                .append(IntelliJPathImpl.fromString(newFileName));
         }
 
-        IPath path = IntelliJPathImpl.fromString(newFile.getPath());
-        IntelliJProjectImpl project = getProjectForResource(path);
+        IActivity activity;
+
+        if (oldPathIsShared && newPathIsShared) {
+            //moved file inside shared modules
+            SPath newFilePath = new SPath(newParentPath.getProject(),
+                newParentPath.getProjectRelativePath().append(relativePath));
+
+            activity = new FileActivity(user, Type.MOVED,
+                FileActivity.Purpose.ACTIVITY, newFilePath, oldFilePath, null,
+                encoding);
+
+            editorManager.replaceAllEditorsForPath(oldFilePath, newFilePath);
+
+            annotationManager.updateAnnotationPath(oldFilePath.getFile(),
+                newFilePath.getFile());
+
+        } else if (newPathIsShared) {
+            //moved file into shared module
+            byte[] fileContent = getContent(oldFile);
+
+            SPath newFilePath = new SPath(newParentPath.getProject(),
+                newParentPath.getProjectRelativePath().append(relativePath));
+
+            activity = new FileActivity(user, Type.CREATED,
+                FileActivity.Purpose.ACTIVITY, newFilePath, null, fileContent,
+                encoding);
+
+            if (fileIsOpen) {
+                Editor editor = projectAPI.openEditor(oldFile, false);
+
+                editorManager.addEditorMapping(newFilePath, editor);
+            }
+
+        } else if (oldPathIsShared) {
+            //moved file out of shared module
+            activity = new FileActivity(user, Type.REMOVED,
+                FileActivity.Purpose.ACTIVITY, oldFilePath, null, null, null);
+
+            editorManager.removeAllEditorsForPath(oldFilePath);
+
+            annotationManager.removeAnnotations(oldFilePath.getFile());
+
+        } else {
+            //neither source nor destination are shared
+            if (LOG.isTraceEnabled()) {
+                LOG.trace(
+                    "Ignoring non-shared resource move - resource: " + oldFile
+                        + ", old base parent: " + oldBaseParent
+                        + ", new base parent: " + newBaseParent);
+            }
 
-        if (project == null || !isValidProject(project) ||
-            !isCompletelyShared(project)) {
             return;
         }
 
-        path = makeAbsolutePathProjectRelative(path, project);
+        fireActivity(activity);
 
-        SPath spath = new SPath(project, path);
+        if (oldPathIsShared) {
+            if (fileIsOpen) {
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Merge this if statement with the enclosing one. ([squid:S1066](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS1066))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=307&issues=AWcYB0JPkEeD8QRk19vP&open=AWcYB0JPkEeD8QRk19vP)

[ISSUE_KEY]: AWcYB0JPkEeD8QRk19vP